src/components/implementation/no_interface/vkernel/vkernel.c

```diff
@@ -0,0 +1,195 @@
+#define PRINT_FN prints
+#define debug_print(str) (PRINT_FN(str __FILE__ ":" STR(__LINE__) ".\n"))
+#define BUG() do { debug_print("BUG @ "); *((int *)0) = 0; } while (0);
+
+struct vms_info {
+	struct cos_compinfo cinfo;
+	captblcap_t ct;
+	pgtblcap_t pt, utpt;
+	compcap_t cc;
+	thdcap_t initthd, exitthd;
+	thdid_t inittid;
+	tcap_t inittcap;
+	arcvcap_t initrcv;
+} vmx_info[COS_VIRT_MACH_COUNT];
+
+struct vkernel_info {
+	struct cos_compinfo cinfo;
+
+	thdcap_t termthd;
+	asndcap_t vminitasnd[COS_VIRT_MACH_COUNT];
+} vk_info;
+
+extern vaddr_t cos_upcall_entry;
+extern void vm_init(void *);
+
+unsigned int ready_vms = COS_VIRT_MACH_COUNT;
+struct cos_compinfo *vk_cinfo = (struct cos_compinfo *)&vk_info.cinfo;
+
+void
+vk_terminate(void *d)
+{
+	BUG();
+}
+
+void
+vm_exit(void *d)
+{
+	printc("%d: EXIT\n", (int)d);
+	ready_vms --;
+	vmx_info[(int)d].initthd = 0;	
+
+	cos_thd_switch(BOOT_CAPTBL_SELF_INITTHD_BASE);
+}
+
+void
+scheduler(void) 
+{
+	static unsigned int i;
+	thdid_t tid;
+	int rcving;
+	cycles_t cycles;
+	int index;
+
+	while (ready_vms) {
+		cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_BASE, &tid, &rcving, &cycles);
+		index = i++ % COS_VIRT_MACH_COUNT;
+		
+		if (vmx_info[index].initthd && vk_info.vminitasnd[index]) {
+			cos_asnd(vk_info.vminitasnd[index]);
+		}
+	}
+}
+
+
+void
+cos_init(void)
+{
+	int id;
+
+	printc("vkernel: START\n");
+	assert(COS_VIRT_MACH_COUNT >= 2);
+
+	cos_meminfo_init(&vk_cinfo->mi, BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ, BOOT_CAPTBL_SELF_UNTYPED_PT);
+	cos_compinfo_init(vk_cinfo, BOOT_CAPTBL_SELF_PT, BOOT_CAPTBL_SELF_CT, BOOT_CAPTBL_SELF_COMP,
+			(vaddr_t)cos_get_heap_ptr(), BOOT_CAPTBL_FREE, vk_cinfo);
+
+	vk_info.termthd = cos_thd_alloc(vk_cinfo, vk_cinfo->comp_cap, vk_terminate, NULL);
+	assert(vk_info.termthd);
+
+	for (id = 0; id < COS_VIRT_MACH_COUNT; id ++) {
+		struct cos_compinfo *vm_cinfo = &vmx_info[id].cinfo;
+		struct vms_info *vm_info = &vmx_info[id];
+		int ret, vm_range, i;
+
+		printc("vkernel: VM%d Init START\n", id);
+		printc("\tForking VM\n");
+		vm_info->exitthd = cos_thd_alloc(vk_cinfo, vk_cinfo->comp_cap, vm_exit, (void *)id);
+		assert(vm_info->exitthd);
+		
+		vm_info->ct = cos_captbl_alloc(vk_cinfo);
+		assert(vm_info->ct);
+
+		vm_info->pt = cos_pgtbl_alloc(vk_cinfo);
+		assert(vm_info->pt);
+
+		vm_info->utpt = cos_pgtbl_alloc(vk_cinfo);
+		assert(vm_info->utpt);
+
+		vm_info->cc = cos_comp_alloc(vk_cinfo, vm_info->ct, vm_info->pt, (vaddr_t)&cos_upcall_entry);
+		assert(vm_info->cc);
+
+		cos_meminfo_init(&vm_cinfo->mi, BOOT_MEM_KM_BASE, COS_VIRT_MACH_UNTYPED_SIZE, vm_info->utpt);
+		cos_compinfo_init(vm_cinfo, vm_info->pt, vm_info->ct, vm_info->cc,
+				(vaddr_t)BOOT_MEM_VM_BASE, VM_CAPTBL_FREE, vk_cinfo);
+
+		vm_info->initthd = cos_thd_alloc(vk_cinfo, vm_cinfo->comp_cap, vm_init, (void *)id);
+		assert(vm_info->initthd);
+		vm_info->inittid = (thdid_t)cos_introspect(vk_cinfo, vm_info->initthd, 9);
```

Comment:
Yeah, there is no defined variable for any of the `thd_introspect()` commands. I'll add them, that should improve readability. Thank you.